common/utils.py

```diff
@@ -19,9 +19,53 @@
 import json
 import re
 from typing import Any
+import datetime
 
 from absl import logging
 from PIL import Image
+import google.auth
+from google.cloud import storage
+from google.auth import impersonated_credentials
+
+import os
+
+def generate_signed_url(gcs_uri: str) -> str:
+    """Generates a signed URL for a GCS object."""
+    if not gcs_uri or not gcs_uri.startswith("gs://"):
+        return ""
+    try:
+        service_account_email = os.environ.get("SERVICE_ACCOUNT_EMAIL")
+        if not service_account_email:
+            logging.error("SERVICE_ACCOUNT_EMAIL environment variable not set.")
+            return ""
+
+        credentials, _ = google.auth.default()
+
+        # Use impersonation to sign the URL.
+        signing_credentials = impersonated_credentials.Credentials(
+            source_credentials=credentials,
+            target_principal=service_account_email,
+            target_scopes=["https://www.googleapis.com/auth/devstorage.read_only"],
+        )
+
+        storage_client = storage.Client(credentials=credentials)
+        bucket_name, blob_name = gcs_uri.replace("gs://", "").split("/", 1)
+        bucket = storage_client.bucket(bucket_name)
+        blob = bucket.blob(blob_name)
+
+        # V4 signed URLs are the recommended and most secure version.
+        signed_url = blob.generate_signed_url(
+            version="v4",
+            expiration=datetime.timedelta(minutes=15),
```

Comment:
Is 15 minutes enough? What happens when the user comes back after 15mins, is there a refresh token?

Reply:
This is a bit of a legacy - we're moving away from signed urls to the fastapi media endpoint proxy. This bit of code is in case I missed some refactors from the old signed url process.

For signed urls, there's no refresh token, it's the time given (in this case 15 min); if someone left their browser open for 16 min and came back, the url wouldn't load, and the page would have to be refreshed. With our /media proxy, this'd be cached in the browser.